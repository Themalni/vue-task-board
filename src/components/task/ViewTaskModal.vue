<template>
  <div>
    <Modal v-if="!isEditing" title="Zadanie" :status="activeTask.status">
      <template v-slot:body>
        <h4 class="mb-1">Opis zadania:</h4>
        <div class="card-grey task-description">
          <p class="text-md m-0 p-1">{{ activeTask.description }}</p>
        </div>
        <h4 class="mt-2 mr-1 section-label">Rodzaj zadania:</h4>
        <Tag class="mt-2" :type="activeTask.type"/>
      </template>
      <template v-slot:footer>
        <Button class="ml-2" color="button-primary" @addAction="isEditing = true">Edytuj</Button>
        <Button class="ml-2" color="button-default" @addAction="closeModal()">Zamknij</Button>
      </template>
    </Modal>
    <Modal title="Edytuj zadanie" :status="activeTask.status" v-else>
      <template v-slot:body>
        <div class="container pb-3">
          <h4>Opis zadania:</h4>
          <textarea
            class="mt-2"
            :class="[{ 'has-error': errors.first('description') }]"
            name="description"
            type="textarea"
            label="Description"
            v-model="activeTask.description"
            v-validate="'required'"
            data-vv-as="Opis zadania"
            placeholder="Dodaj opis zadania..."
            rows="6"
            @focus="isEditing = true"
          ></textarea>
          <small
            v-if="errors.first('description')"
            class="text-xs error-text edit-error"
          >{{ errors.first('description') }}</small>
        </div>
        <div class="container pb-3">
          <h4 class="w-100">Rodzaj zadania:</h4>
          <Select
            :name="'taskTypesSelect'"
            id="select"
            label
            v-model="activeTask.type"
            rules="required"
            friendly-name="Rodzaj zadania"
            :options="taskTypes"
            default-option="Wybierz rodzaj zadania"
          ></Select>
        </div>
      </template>
      <template v-slot:footer>
        <Button class="button-left" color="button-danger" :disabled="!isEditing" @addAction="openDeleteTaskModal()">Usuń</Button>
        <Button color="button-primary" :disabled="!isEditing" @addAction="setChanges(activeTask)">Zachowaj</Button>
        <Button class="ml-2" color="button-default" @addAction="closeModal()">Anuluj</Button>
      </template>
    </Modal>
    <Modal v-if="deleteModalIsOpen" title="Podtwierdzenie usuwania zadania" :small="true">
      <template v-slot:body>
        <h4>Czy napewno chcesz usunąć to zadanie?</h4>
      </template>
      <template v-slot:footer>
        <Button color="button-danger" :disabled="!isEditing" @addAction="deleteSelectedTask(activeTask)">Usuń</Button>
        <Button class="ml-2" color="button-default" @addAction="closeDeleteModal()">Anuluj</Button>
      </template>
    </Modal>
  </div>
</template>

<script>
import Modal from '@/components/Modal'
import Button from '@/elements/Button'
import Select from '@/elements/Select'
import Tag from '@/elements/Tag'
import { mapActions, mapGetters } from 'vuex'

export default {
  name: 'ViewTaskModal',
  inject: ['$validator'],
  components: {
    Modal,
    Button,
    Select,
    Tag
  },
  data: () => ({
    isEditing: false,
    deleteModalIsOpen: false,
    taskTypes: {
      feature: 'feature',
      bugfix: 'bugfix',
      default: 'default'
    }
  }),
  computed: {
    ...mapGetters(['activeTask', 'editModalIsOpen'])
  },
  methods: {
    ...mapActions([
      'changeViewTaskModalState',
      'changeNewTaskModalState',
      'saveTask',
      'saveChanges',
      'deleteTask'
    ]),
    closeModal () {
      this.$store.dispatch('changeViewTaskModalState', false)
    },
    closeDeleteModal () {
      this.deleteModalIsOpen = false
    },
    setChanges (taskToSave) {
      this.$store.dispatch('saveChanges', taskToSave)
      this.$store.dispatch('changeViewTaskModalState', false)
    },
    openNewTaskModal () {
      this.$store.dispatch('changeViewTaskModalState', false)
      this.$store.dispatch('changeNewTaskModalState', true)
    },
    openDeleteTaskModal () {
      this.deleteModalIsOpen = true
    },
    deleteSelectedTask (task) {
      this.$store.dispatch('deleteTask', task)
      this.deleteModalIsOpen = false
      this.$store.dispatch('changeViewTaskModalState', false)
    }
  }
}
</script>

<style lang="scss">
.separator {
  height: 1px;
  width: 100%;
  border: none;
  background-color: $grey;
}
.task-description {
  min-height: 80px;
  width: 100%;
}
.section-label {
  line-height: 0.7;
}
.button-left {
  position: absolute;
  left: 1.5em;
}
</style>
