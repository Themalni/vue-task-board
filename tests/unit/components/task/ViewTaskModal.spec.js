import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import VeeValidate from 'vee-validate'
import mockState from '../../utils/mockState'
import ViewTaskModal from '@/components/task/ViewTaskModal'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VeeValidate)

describe('ViewTaskModal', () => {
  let wrapper
  let store
  let actions
  let state
  let getters

  beforeEach(() => {
    state = mockState.state
    getters = {
      activeTask: () => state.categories[0].tasks[0],
      editModalIsOpen: () => false
    }
    actions = {
      changeViewTaskModalState: jest.fn(),
      changeNewTaskModalState: jest.fn(),
      saveTask: jest.fn(),
      saveChanges: jest.fn(),
      deleteTask: jest.fn()
    }
    store = new Vuex.Store({
      state,
      getters,
      actions
    })

    wrapper = mount(ViewTaskModal, {
      setData: ({
        isEditing: false,
        deleteModalIsOpen: false,
        taskTypes: {
          feature: 'feature',
          bugfix: 'bugfix',
          default: 'default'
        }
      }),
      store,
      localVue,
      sync: false
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('calls `closeModal` when button is clicked', () => {
    wrapper.vm.closeModal()
    expect(actions.changeViewTaskModalState.mock.calls).toHaveLength(1)
  })
  it('calls `openDeleteTaskModal` when button is clicked', () => {
    wrapper.vm.openDeleteTaskModal()
    expect(wrapper.vm.deleteModalIsOpen).toBeTruthy()
  })
  it('calls `closeDeleteModal` when button is clicked', () => {
    wrapper.vm.closeDeleteModal()
    expect(wrapper.vm.deleteModalIsOpen).toBeFalsy()
  })
  it('calls `setChanges` when button is clicked', () => {
    wrapper.vm.setChanges()
    expect(actions.saveChanges.mock.calls).toHaveLength(1)
    expect(actions.changeViewTaskModalState.mock.calls).toHaveLength(1)
  })
  it('calls `openNewTaskModal` when button is clicked', () => {
    wrapper.vm.openNewTaskModal()
    expect(actions.changeViewTaskModalState.mock.calls).toHaveLength(1)
    expect(actions.changeNewTaskModalState.mock.calls).toHaveLength(1)
  })
  it('calls `deleteSelectedTask` when button is clicked', () => {
    wrapper.vm.deleteSelectedTask()
    expect(actions.deleteTask.mock.calls).toHaveLength(1)
    expect(actions.changeViewTaskModalState.mock.calls).toHaveLength(1)
    expect(wrapper.vm.deleteModalIsOpen).toBeFalsy()
  })
})
