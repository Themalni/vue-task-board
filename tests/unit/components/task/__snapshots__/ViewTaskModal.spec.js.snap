// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewTaskModal renders correctly 1`] = `
<div>
  <div class="overlay" name="fade">
    <div class="modal">
      <div class="modal-header">
        <h3 class="modal-header-text text-white">Zadanie</h3>
        <h3 class="modal-header-text text-white ml-1">(OczekujÄ…ce)</h3>
      </div>
      <div class="modal-body">
        <h4 class="mb-1">Opis zadania:</h4>
        <div class="card-grey task-description">
          <p class="text-md m-0 p-1">Opis zadania #1</p>
        </div>
        <h4 class="mt-2 mr-1 section-label">Rodzaj zadania:</h4>
        <div class="mt-2 tag text-sm tag-feature">
          feature
        </div>
      </div>
      <div class="modal-footer">
        <div class="ml-2"><button class="button button-primary">Edytuj</button></div>
        <div class="ml-2"><button class="button button-default">Zamknij</button></div>
      </div>
    </div>
  </div>
  <!---->
</div>
`;
